modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequestHandler.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.http.netty4;
+
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+import io.netty.handler.codec.http.FullHttpRequest;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.http.netty4.pipelining.HttpPipelinedRequest;
+
+@ChannelHandler.Sharable
+class Netty4HttpRequestHandler extends SimpleChannelInboundHandler<Object> {
+
+    private final Netty4HttpServerTransport serverTransport;
+    private final boolean httpPipeliningEnabled;
+    private final boolean detailedErrorsEnabled;
+    private final ThreadContext threadContext;
+
+    Netty4HttpRequestHandler(Netty4HttpServerTransport serverTransport, boolean detailedErrorsEnabled, ThreadContext threadContext) {
+        this.serverTransport = serverTransport;
+        this.httpPipeliningEnabled = serverTransport.pipelining;
+        this.detailedErrorsEnabled = detailedErrorsEnabled;
+        this.threadContext = threadContext;
+    }
+
+    @Override
+    protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Exception {
+        final FullHttpRequest request;
+        final HttpPipelinedRequest pipelinedRequest;
+        if (this.httpPipeliningEnabled && msg instanceof HttpPipelinedRequest) {
+            pipelinedRequest = (HttpPipelinedRequest) msg;
+            request = (FullHttpRequest) pipelinedRequest.last();
+        } else {
+            pipelinedRequest = null;
+            request = (FullHttpRequest) msg;
+        }
+
+        threadContext.copyHeaders(request.headers());
```

Comment:
Contained in 0878d3b582978834309b42197088314b97c90ae9 although I meant for it to be a separate push.